### PR TITLE
ABI functions return WasmResult.

### PR DESCRIPTION
**Pull request by jplevyak**
For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

File: api/wasm/cpp/proxy_wasm_result.h [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  NotFound = 1,
+  BadType = 2,
+  BadArgument = 3,
+  SerializationFailure = 4,
+  ParseFailure = 5,
+  NoSuchStruct = 6,
+  NoSuchField = 7,
+  BadExpression = 8,
+  MemoryError = 9,
+  Eos = 10,
```
Comment (PiotrSikora) [inline api/wasm/cpp/proxy_wasm_result.h]: Erm, I really don't like this being so short, but also, `end-of-stream` in proxies is mostly used in context of real stream, similar to `EOF`, but the way this is used now seems to have the meaning "no more entries in the queue", which could probably be represented by `WasmResult::Ok` + empty response.

File: api/wasm/cpp/proxy_wasm_result.h [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+enum class WasmResult : uint32_t {
+  Ok = 0,
+  NotFound = 1,
+  BadType = 2,
+  BadArgument = 3,
+  SerializationFailure = 4,
+  ParseFailure = 5,
+  NoSuchStruct = 6,
+  NoSuchField = 7,
+  BadExpression = 8,
```
[the remark this answers is not in the document]
Comment (jplevyak) [reply api/wasm/cpp/proxy_wasm_result.h]: NotFound

File: api/wasm/cpp/proxy_wasm_result.h (hunk 1 above)
Comment (jplevyak) [reply api/wasm/cpp/proxy_wasm_result.h]: Empty

File: api/wasm/cpp/proxy_wasm_result.h [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  MemoryError = 9,
```
[the remark this answers is not in the document]
Comment (jplevyak) [reply api/wasm/cpp/proxy_wasm_result.h]: SGTM

File: api/wasm/cpp/proxy_wasm_result.h [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+#include <string>
+
+enum class WasmResult : uint32_t {
+  Ok = 0,
+  NotFound = 1,
+  BadType = 2,
+  BadArgument = 3,
+  SerializationFailure = 4,
+  ParseFailure = 5,
+  NoSuchStruct = 6,
```
[the remark this answers is not in the document]
Comment (jplevyak) [reply api/wasm/cpp/proxy_wasm_result.h]: NotFound

File: api/wasm/cpp/proxy_wasm_result.h [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+#pragma once
+
+#include <string>
+
+enum class WasmResult : uint32_t {
+  Ok = 0,
+  NotFound = 1,
+  BadType = 2,
+  BadArgument = 3,
```
[the remark this answers is not in the document]
Comment (jplevyak) [reply api/wasm/cpp/proxy_wasm_result.h]: BadArgument it is

File: api/wasm/cpp/proxy_wasm_api.h [hunk 6]
```diff
@@ -24,25 +24,36 @@ class ProxyException : std::runtime_error {
 };
 #endif
 
-inline void logTrace(const std::string& logMessage) {
-  proxy_log(LogLevel::trace, logMessage.c_str(), logMessage.size());
+#define CHECK_WASM_RESULT(_r) do { \
+  auto _result = _r; \
+  (void) _result; /* unused */ \
+  assert(_result == WasmResult::Ok); \
```
[the remark this answers is not in the document]
Comment (jplevyak) [reply api/wasm/cpp/proxy_wasm_api.h]: assert only works in when NDEBUG is not defined.

File: source/extensions/common/wasm/v8/v8.cc [hunk 7]
```diff
@@ -523,36 +494,44 @@ uint64_t V8::getMemorySize() {
...
 }
 
-bool V8::setWord(uint64_t pointer, uint64_t word) {
-  ENVOY_LOG(trace, "[wasm] setWord({}, {})", pointer, word);
-  ASSERT(memory_ != nullptr);
-  RELEASE_ASSERT(pointer + sizeof(uint32_t) <= memory_->data_size(), "");
-  uint32_t word32 = static_cast<uint32_t>(word);
-  ::memcpy(memory_->data() + pointer, &word32, sizeof(uint32_t));
+bool V8::setWord(uint64_t pointer, Word word) {
+  ENVOY_LOG(trace, "[wasm] setWord({}, {})", word, pointer);
```
[the remark this answers is not in the document]
Comment (jplevyak) [reply source/extensions/common/wasm/v8/v8.cc]: Fixed.

File: api/wasm/cpp/proxy_wasm_api.h [hunk 8]
```diff
@@ -24,25 +24,36 @@ class ProxyException : std::runtime_error {
 };
 #endif
 
-inline void logTrace(const std::string& logMessage) {
-  proxy_log(LogLevel::trace, logMessage.c_str(), logMessage.size());
+#define CHECK_WASM_RESULT(_r) do { \
```
[the remark this answers is not in the document]
Comment (jplevyak) [reply api/wasm/cpp/proxy_wasm_api.h]: done

File: source/extensions/common/wasm/null/wasm_api_impl.h [hunk 9]
```diff
@@ -19,134 +19,156 @@ namespace Plugin {
 #define WS(_x) Word(static_cast<uint64_t>(_x))
 #define WR(_x) Word(reinterpret_cast<uint64_t>(_x))
 
+inline WasmResult toWasmResult(Word w) { return static_cast<WasmResult>(w.u64); }
```
[the remark this answers is not in the document]
Comment (jplevyak) [reply source/extensions/common/wasm/null/wasm_api_impl.h]: What, you don't like function overloading?  Sure.  Done.

File: api/wasm/cpp/proxy_wasm_api.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (jplevyak) [reply api/wasm/cpp/proxy_wasm_api.h]: I want this to occur even if NDEBUG is set, so "assert" seems to be the wrong word.  Combining them as CHECK_OK.  WDYT?

[the remark this answers is not in the document]
Comment (jplevyak) [reply api/wasm/cpp/proxy_wasm_api.h]: Make that CHECK_RESULT

File: api/wasm/cpp/proxy_wasm_api.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (PiotrSikora) [reply api/wasm/cpp/proxy_wasm_api.h]: I understand that part, but I don't get the point of `auto _result = _r;`.

File: api/wasm/cpp/proxy_wasm_api.h [hunk 10]
```diff
@@ -8,6 +8,9 @@
 #include <utility>
 #include <vector>
 
+#define CHECK_RESULT(_c) do { if ((_c) != WasmResult::Ok) { proxy_log(LogLevel::critical, #_c, sizeof( #_c )-1); abort(); }} while(0)
+#
```
Comment (PiotrSikora) [inline api/wasm/cpp/proxy_wasm_api.h]: Nit: empty `#`.

File: api/wasm/cpp/proxy_wasm_api.h [hunk 11]
```diff
@@ -24,25 +27,30 @@ class ProxyException : std::runtime_error {
...
+  return proxy_log(LogLevel::critical, logMessage.data(), logMessage.size());
 }
-inline void logCritical(const std::string& logMessage) {
-  proxy_log(LogLevel::critical, logMessage.c_str(), logMessage.size());
+inline void logAbort(StringView logMessag) {
+  logCritical(logMessag);
+  abort();
 }
 
+
```
Comment (PiotrSikora) [inline api/wasm/cpp/proxy_wasm_api.h]: Nit: extra empty line.

File: api/wasm/cpp/proxy_wasm_result.h [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+  SerializationFailure = 4,
+  ParseFailure = 5,
+  NoSuchStruct = 6,
+  NoSuchField = 7,
```
[the remark this answers is not in the document]
Comment (jplevyak) [reply api/wasm/cpp/proxy_wasm_result.h]: Done

File: api/wasm/cpp/proxy_wasm_api.h (hunk 10 above)
Comment (jplevyak) [reply api/wasm/cpp/proxy_wasm_api.h]: Done

File: api/wasm/cpp/proxy_wasm_api.h (hunk 11 above)
Comment (jplevyak) [reply api/wasm/cpp/proxy_wasm_api.h]: Done

File: api/wasm/cpp/proxy_wasm_api.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (jplevyak) [reply api/wasm/cpp/proxy_wasm_api.h]: So that the call occurs even if NDEBUG is not defined.  In any case I removed this code.

File: api/wasm/cpp/Makefile.base [hunk 13]
```diff
@@ -6,10 +6,9 @@ else
   CPP_CONTEXT_LIB = ${CPP_API}/proxy_wasm_intrinsics.cc
 endif
 
-%.wasm %.wat: %.cc ${CPP_API}/proxy_wasm_intrinsics.h ${CPP_API}/proxy_wasm_enums.h ${CPP_API}/proxy_wasm_externs.h ${CPP_API}/proxy_wasm_impl.h ${CPP_API}/proxy_wasm_intrinsics.js ${CPP_CONTEXT_LIB}
+%.wasm %.wat: %.cc ${CPP_API}/proxy_wasm_intrinsics.h ${CPP_API}/proxy_wasm_enums.h ${CPP_API}/proxy_wasm_externs.h ${CPP_API}/proxy_wasm_api.h ${CPP_API}/proxy_wasm_intrinsics.js ${CPP_CONTEXT_LIB}
 	em++ -s WASM=1 -s BINARYEN_TRAP_MODE='clamp' -s LEGALIZE_JS_FFI=0 -s EMIT_EMSCRIPTEN_METADATA=1 --std=c++17 -O3 -g3 -I${CPP_API} -I/usr/local/include --js-library ${CPP_API}/proxy_wasm_intrinsics.js $*.cc ${CPP_API}/proxy_wasm_intrinsics.pb.cc ${CPP_CONTEXT_LIB} ${CPP_API}/libprotobuf.bc -o $*.js
 	wavm-disas $*.wasm $*.wat
-	wavm-compile $*.wasm $*.wasm
```
Comment (PiotrSikora) [inline api/wasm/cpp/Makefile.base]: Could you split to this another PR? Ideally, also removing `allow_precompiled` from the codebase (assuming that's the intention here) and relevant tests for it? There is also one more use of `wavm-compile` in `test/extensions/wasm/test_data/Makefile`.

File: api/wasm/cpp/proxy_wasm_externs.h [hunk 14]
```diff
@@ -128,17 +132,16 @@ extern "C" uint32_t proxy_grpcCall(const char* service_ptr, size_t service_size,
...
-extern "C" void proxy_incrementMetric(uint32_t metric_id, int64_t offset);
-extern "C" void proxy_recordMetric(uint32_t metric_id, uint64_t value);
-extern "C" uint64_t proxy_getMetric(uint32_t metric_id);
+extern "C" WasmResult proxy_defineMetric(MetricType type, const char* name_ptr, size_t name_size, uint32_t* metric_id);
+extern "C" WasmResult proxy_incrementMetric(uint32_t metric_id, int64_t offset);
+extern "C" WasmResult proxy_recordMetric(uint32_t metric_id, uint64_t value);
+extern "C" WasmResult proxy_getMetric(uint32_t metric_id, uint64_t* result);
 
 // System
 // Returns 0 on success.
```
Comment (PiotrSikora) [inline api/wasm/cpp/proxy_wasm_externs.h]: Nit: fix comment.

File: api/wasm/cpp/proxy_wasm_externs.h [hunk 15]
```diff
@@ -52,67 +52,71 @@
...
 // Consequently it should typically be followed by a call to proxy_dequeueSharedQueue.
-extern "C" uint32_t proxy_registerSharedQueue(const char* queue_name_ptr, size_t queue_name_size);
-extern "C" uint32_t proxy_resolveSharedQueue(const char* vm_id, size_t vm_id_size, const char* queue_name_ptr, size_t queue_name_size);
-// Returns true on end-of-stream (no more data available).
-extern "C" bool proxy_dequeueSharedQueue(uint32_t token, const char** data_ptr, size_t* data_size);
+// Returns: Ok
+extern "C" WasmResult proxy_registerSharedQueue(const char* queue_name_ptr, size_t queue_name_size, uint32_t* token);
+// Returns: Ok, NotFound
+extern "C" WasmResult proxy_resolveSharedQueue(const char* vm_id, size_t vm_id_size, const char* queue_name_ptr, size_t queue_name_size, uint32_t* token);
+// Returns Ok, Eos, NotFound (token not registered).
```
Comment (PiotrSikora) [inline api/wasm/cpp/proxy_wasm_externs.h]: Nit: `Eos` -> `Empty`.

File: api/wasm/cpp/Makefile.base (hunk 13 above)
Comment (jplevyak) [reply api/wasm/cpp/Makefile.base]: https://github.com/envoyproxy/envoy-wasm/pull/135

Comment (jplevyak) [reply api/wasm/cpp/Makefile.base]: I think we do want to allow precompiling since that could be a big win for WAVM.   However it will come with some warnings about using the right version of WAVM.

File: api/wasm/cpp/proxy_wasm_externs.h (hunk 14 above)
Comment (jplevyak) [reply api/wasm/cpp/proxy_wasm_externs.h]: Done.

File: api/wasm/cpp/proxy_wasm_externs.h (hunk 15 above)
Comment (jplevyak) [reply api/wasm/cpp/proxy_wasm_externs.h]: Done.

**Comment by jplevyak** [comment]
Removed those files.

File: test/extensions/filters/http/wasm/test_data/metadata_cpp.cc [hunk 16]
```diff
@@ -22,19 +22,19 @@ static RegisterContextFactory register_ExampleContext(CONTEXT_FACTORY(ExampleCon
 
 void ExampleRootContext::onTick() {
   google::protobuf::Value value;
-  if (MetadataResult::Ok != nodeMetadataValue("wasm_node_get_key", &value)) {
-    logDebug("onTick metadata error");
+  if (auto r = nodeMetadataValue("wasm_node_get_key", &value); r != WasmResult::Ok) {
```
Comment (PiotrSikora) [inline test/extensions/filters/http/wasm/test_data/metadata_cpp.cc]: Nit: This isn't Go, could you split this into 2 lines?

Comment (jplevyak) [reply test/extensions/filters/http/wasm/test_data/metadata_cpp.cc]: Done.

You don't like the new c++17 syntax?  It doesn't leave the WasmResult hanging in the surrounding scope.

**Comment by jplevyak** [comment]
I don't see anything unresolved.

**Comment by jplevyak** [comment]
Found them "hidden"

File: source/extensions/common/wasm/wasm.cc [hunk 17]
```diff
@@ -45,18 +45,16 @@ namespace Common {
...
-thread_local Context* current_context_ = nullptr;
 #define WASM_CONTEXT(_c)                                                                           \
   (ContextOrEffectiveContext(static_cast<Context*>((void)_c, current_context_)))
 // The id of the context which should be used for calls out of the VM in place of current_context_
 // above.
-thread_local uint32_t effective_context_id_ = 0; // 0 indicates no effective context id.
 
 namespace {
 
+Word toWord(WasmResult r) { return Word(static_cast<uint64_t>(r)); }
```
[the remark this answers is not in the document]
Comment (jplevyak) [reply source/extensions/common/wasm/wasm.cc]: done

File: api/wasm/cpp/proxy_wasm_result.h [hunk 18]
```diff
@@ -0,0 +1,40 @@
+/*
+ * Intrinsic enumerations available to WASM modules.
+ */
+// NOLINT(namespace-envoy)
+
+#pragma once
+
+#include <string>
+
+enum class WasmResult : uint32_t {
```
[the remark this answers is not in the document]
Comment (jplevyak) [reply api/wasm/cpp/proxy_wasm_result.h]: done

File: test/extensions/filters/http/wasm/test_data/metadata_cpp.cc (hunk 16 above)
Comment (PiotrSikora) [reply test/extensions/filters/http/wasm/test_data/metadata_cpp.cc]: Envoy is still using C++14.
